src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs

```diff
@@ -4607,5 +4606,158 @@ public static void Main()
                 Diagnostic(ErrorCode.ERR_AssignmentInitOnly, "Property").WithArguments("ITest28.Property").WithLocation(9, 47)
                 );
         }
+
+        [Fact, WorkItem(50696, "https://github.com/dotnet/roslyn/issues/50696")]
+        public void PickAmbiguousAttributeFromCorlib()
+        {
+            var corlib_cs = @"
+namespace System
+{
+    public class Object { }
+    public struct Int32 { }
+    public struct Boolean { }
+    public class String { }
+    public class ValueType { }
+    public struct Void { }
+    public class Attribute { }
+}
+";
+
+            string source = @"
+public class C
+{
+    public int Property { get; init; }
+}
+";
+            var corlibWithoutIsExternalInitRef = CreateEmptyCompilation(corlib_cs, assemblyName: "corlibWithoutIsExternalInit")
+                .EmitToImageReference();
+
+            var corlibWithIsExternalInitRef = CreateEmptyCompilation(corlib_cs + IsExternalInitTypeDefinition, assemblyName: "corlibWithIsExternalInit")
+                .EmitToImageReference();
+
+            var libWithIsExternalInitRef = CreateEmptyCompilation(IsExternalInitTypeDefinition, references: new[] { corlibWithoutIsExternalInitRef }, assemblyName: "libWithIsExternalInit")
+                .EmitToImageReference();
+
+            var libWithIsExternalInitRef2 = CreateEmptyCompilation(IsExternalInitTypeDefinition, references: new[] { corlibWithoutIsExternalInitRef }, assemblyName: "libWithIsExternalInit2")
+                .EmitToImageReference();
+
+            {
+                // attribute in source
+                var comp = CreateEmptyCompilation(new[] { source, IsExternalInitTypeDefinition }, references: new[] { corlibWithoutIsExternalInitRef });
+                comp.VerifyEmitDiagnostics();
+                var modifier = getUsedModifier(comp);
+                Assert.True(modifier.Modifier.IsInSource());
```

Comment:
>IsInSource [](start = 46, length = 10)

Relying on IsInSource helper here feels fragile, we are relying on an assumption that all references are metadata references. I think it would be better to verify that the type belongs to the source module of the compilation.